### PR TITLE
rfc: Use `django-manifest-loader` for static assets

### PR DESCRIPTION
**Pull request by billyvg**
This changes the way we handle static assets. Previously, the assets were versioned by either a `version` file (commit sha?) or a timestamp. This changes it so that the `version` is removed from the path and instead change webpack to include a hash (based on file contents) in the filename. Webpack also produces a manifest.json file that maps a key -> hashed filename which `django-manifest-loader` uses to load static assets.

Old:

```
/_static/{version}/sentry/dist/app.js
```

New:
```
/_static/sentry/dist/app.{appHash}.js
```

* [x] Check CDN (fastly) settings to make sure that non-webpack assets are not cached long term

~After thinking about this some more, I don't think we'll need this long-term as we will be aiming to have a webpack runtime chunk/bootloader handle loading all webpack assets (so remove loading webpack assets from django views), but this may be a good intermediate step if we wanted to evaluate CDN caching.~

~We will need to refactor `sentry.css` so that it is split into CSS that's needed during initial rendering state and application level css which is imported via js application (vs being included in server template).~

### CODE REVIEW

File: src/sentry/utils/assets.py [hunk 1]
```diff
@@ -1,5 +1,22 @@
 from django.conf import settings
 
+from manifest_loader.utils import _get_manifest, _load_from_manifest
```
Comment (billyvg) [inline src/sentry/utils/assets.py:3]: 😬

File: webpack.config.js [hunk 2]
```diff
@@ -273,7 +273,7 @@ let appConfig = {
             options: {
               // This needs to be `false` because of platformicons package
               esModule: false,
-              name: '[name].[hash:6].[ext]',
+              name: '[folder]/[name].[hash:6].[ext]',
```
Comment (billyvg) [inline webpack.config.js:276]: Gonna pull this out into a sep pr

File: src/sentry/utils/assets.py [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+def get_manfifest_url(module, key):
+    """
+    Returns an asset URL that is produced by webpack. Uses webpack's manifest to map
+    `key` to the asset produced by webpack. Required if using file contents based hashing for filenames.
+
+    Example:
+      {% manifest_asset_url 'sentry' 'sentry.css' %}
+      =>  "/_static/sentry/dist/sentry.filehash123.css"
+    """
+    manifest_obj = _get_manifest()
```
Comment (joshuarli) [inline src/sentry/utils/assets.py:15]: I know this is WIP, but for expensive (presumably this reads the webpack manifest file) function calls like this, you should wrap it with an unbounded function call cache: `@functools.lru_cache(maxsize=None)`. Unbounded is good here because we know it'll only ever be 1 cached call (no arguments), and it doesn't need to evict so it's faster than maxsize=N (there's a fast path for this in CPython).

File: src/sentry/utils/assets.py [hunk 4]
```diff
@@ -1,5 +1,22 @@
 from django.conf import settings
 
+from manifest_loader.utils import _get_manifest, _load_from_manifest
+
+
+def get_manfifest_url(module, key):
```
Comment (joshuarli) [inline src/sentry/utils/assets.py:6]: Typo here.

Comment (joshuarli) [reply src/sentry/utils/assets.py:6]: Also could just slap a bounded lru_cache on this.

File: src/sentry/templatetags/sentry_assets.py [hunk 5]
```diff
@@ -5,12 +5,13 @@
 from django.utils.safestring import mark_safe
 
 from sentry import options
-from sentry.utils.assets import get_asset_url
+from sentry.utils.assets import get_asset_url, get_manfifest_url
 from sentry.utils.http import absolute_uri
 
 register = template.Library()
 
 register.simple_tag(get_asset_url, name="asset_url")
```
Comment (joshuarli) [inline src/sentry/templatetags/sentry_assets.py:13]: Wouldn't we need to remove all usage of asset_url and get_asset_url? Unless you plan on building both for now, which... would be a lot.

File: src/sentry/utils/assets.py (hunk 3 above)
Comment (billyvg) [reply src/sentry/utils/assets.py:15]: @joshuarli so the manifest library seems to cache this: https://github.com/shonin/django-manifest-loader/blob/dev/manifest_loader/utils.py#L62

File: src/sentry/templatetags/sentry_assets.py (hunk 5 above)
Comment (billyvg) [reply src/sentry/templatetags/sentry_assets.py:13]: We will need to keep `asset_url` as there are files that webpack does not touch, I think ideally we'd have these assets go through webpack too

Comment (billyvg) [reply src/sentry/templatetags/sentry_assets.py:13]: MM I may do a pass to pipe these assets through webpack

File: src/sentry/utils/assets.py (hunk 3 above)
Comment (joshuarli) [reply src/sentry/utils/assets.py:15]: Oh, that's good.

File: src/sentry/testutils/cases.py [hunk 6]
```diff
@@ -132,7 +132,7 @@ def static_asset_manifest(cls, manifest_data):
             finally:
                 # Instead of unlinking, preserve an empty manifest file so that other tests that
                 # may or may not load static assets, do not fail
-                json.dump(manifest_data, {})
```
Comment (joshuarli) [inline src/sentry/testutils/cases.py]: 🅱️ ad 🅱️ illy

File: src/sentry/static/sentry/app/index.tsx [hunk 7]
```diff
@@ -6,37 +6,36 @@ import {Config} from 'app/types';
 
 const BOOTSTRAP_URL = '/api/client-config/';
 
+function renderApp() {
+  if (document.readyState === 'complete') {
+    window.SentryRenderApp();
+  } else {
+    document.addEventListener('DOMContentLoaded', () => window.SentryRenderApp());
+  }
+}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/index.tsx:15]: @EvanPurkhiser Do you see any issues with this? I'm planning on moving this into a separate PR but with some changes: we don't need to attach `SentryRenderApp` to `window` - I think we can move it into a separate module from `bootstrap` where we can then import + render.

Thoughts?

Comment (billyvg) [reply src/sentry/static/sentry/app/index.tsx:15]: Hmm actually, this may interfere with @scefali's Pipeline rendering...

Comment (billyvg) [reply src/sentry/static/sentry/app/index.tsx:15]: @scefali can we potentially move the pipeline rendering into its own webpack entry? how much of sentry's bootstrapping does it need?

Comment (scefali) [reply src/sentry/static/sentry/app/index.tsx:15]: @billyvg Another webpack entry for the pipeline makes sense. As far as bootstrapping goes, it only uses `RenderPipelineView`.

File: src/sentry/templates/sentry/layout.html [hunk 8]
```diff
@@ -43,10 +44,13 @@
 
   {% block scripts %}
   {% locale_js_include %}
-  {% asset_url "sentry" "dist/vendor.js" as asset_url %}
+  {% manifest_asset_url "sentry" "runtime.js" as asset_url %}
```
Comment (billyvg) [inline src/sentry/templates/sentry/layout.html:47]: Might want to rename this `webpack_asset_url`?

File: src/sentry/web/frontend/generic.py [hunk 9]
```diff
@@ -33,13 +33,29 @@ def resolve(path):
     return os.path.split(absolute_path)
 
 
+def static_media_with_manifest(request, **kwargs):
```
Comment (billyvg) [inline src/sentry/web/frontend/generic.py:36]: This is largely unnecessary, we could move this into `static_media`, but in this case we lean on django URL routing a bit to parse the URL.
